pkg/validation/validate.go

```diff
@@ -128,32 +129,108 @@ func ValidateLabels(limits LabelValidationLimits, tenantID string, ls []*typesv1
-		var ok bool
-		if origName, l.Name, ok = SanitizeLabelName(l.Name); !ok {
+		if origName, newName, ok := SanitizeLabelName(l.Name); ok && origName != newName {
+			var err error
+			ls, idx, err = handleSanitizedLabel(ls, idx, origName, newName)
+			if err != nil {
+				return err
+			}
+			lastLabelName = ""
+			if idx > 0 && idx <= len(ls) {
+				lastLabelName = ls[idx-1].Name
+			}
+			continue
+		} else if !ok {
 			return NewErrorf(InvalidLabels, InvalidLabelsErrorMsg, phlaremodel.LabelPairsString(ls), "invalid label name '"+origName+"'")
 		}
 		if !model.LabelValue(l.Value).IsValid() {
 			return NewErrorf(InvalidLabels, InvalidLabelsErrorMsg, phlaremodel.LabelPairsString(ls), "invalid label value '"+l.Value+"'")
 		}
 		if cmp := strings.Compare(lastLabelName, l.Name); cmp == 0 {
-			return NewErrorf(DuplicateLabelNames, DuplicateLabelNamesErrorMsg, phlaremodel.LabelPairsString(ls), origName)
+			return NewErrorf(DuplicateLabelNames, DuplicateLabelNamesErrorMsg, phlaremodel.LabelPairsString(ls), l.Name)
 		}
 		lastLabelName = l.Name
+		idx += 1
 	}
 
 	return nil
 }
 
+// handleSanitizedLabel handles the case where a label name is sanitized. It ensures that the label name is unique and fails if the value is distinct.
+func handleSanitizedLabel(ls []*typesv1.LabelPair, origIdx int, origName, newName string) ([]*typesv1.LabelPair, int, error) {
+	origValue := ls[origIdx].Value
+	// check if sanitized name already exists
+	for i := 0; i < len(ls); i++ {
+		if i == origIdx {
+			continue
+		}
+		if ls[i].Name == newName {
+			if ls[i].Value == origValue {
+				// Same name and value after sanitization - remove the original and continue
+				// Shift all labels after origIdx down by one
+				copy(ls[origIdx:], ls[origIdx+1:])
+				return ls[:len(ls)-1], origIdx, nil
+			} else {
+				return ls, origIdx, NewErrorf(DuplicateLabelNames, DuplicateLabelNamesAfterSanitizationErrorMsg, phlaremodel.LabelPairsString(ls), newName, origName)
+			}
+		}
+	}
+
+	ls[origIdx].Name = newName
+
+	insertIdx := origIdx
+
+	// Find the correct position for the sanitized label. Move backwards to find where it should be inserted
+	for i := 0; i < origIdx; i++ {
+		if newName < ls[i].Name {
+			insertIdx = i
+			break
+		}
+	}
+
+	// If we need to move it forward, find the correct position
+	if insertIdx == origIdx {
+		for i := origIdx + 1; i < len(ls); i++ {
+			if newName > ls[i].Name {
+				insertIdx = i
+			} else {
+				break
+			}
+		}
+	}
+
+	// If position changed, move the label to the correct position
+	if insertIdx != origIdx {
+		label := ls[origIdx]
+		if insertIdx < origIdx {
+			// Move backwards - shift labels right
+			copy(ls[insertIdx+1:origIdx+1], ls[insertIdx:origIdx])
+			ls[insertIdx] = label
+			return ls, insertIdx, nil
+		} else {
+			// Move forwards - shift labels left
+			copy(ls[origIdx:insertIdx], ls[origIdx+1:insertIdx+1])
+			ls[insertIdx] = label
+			return ls, origIdx, nil
+		}
+	}
+
+	return ls, origIdx, nil
+}
```

Comment:
This is also might fault as I started this bit, but I think we a trying to much here and it gets quite complex. I think we can use some stdlib to make this easier:

* We copy the whole slice into a new without origIdx
* We use `slices.BinarySearchFunc` to find if a `newName` is already existing or if not at which position it is
* If its existing we return like the code does right now
* If not then `slices.Insert` the new one at the right place.

This will be less relying on our implementation, like more performant and easier to read.

Let me know if you have any questions.